### PR TITLE
Certificate and oAuth client listing optimization

### PR DESCRIPTION
**Pull request by tbuchaillot**
Related [#1740 tyk-analytics](https://github.com/TykTechnologies/tyk-analytics/issues/1740)

The idea behind this change is to create an index list with the keys values of certs / OAuth clients to avoid SCAN or KEYS operations.

### CODE REVIEW

File: certs/manager.go [hunk 1]
```diff
@@ -488,10 +501,18 @@ func (c *CertificateManager) Add(certData []byte, orgID string) (string, error)
...
 	}
 
+	c.storage.AppendToSet(orgID+"-index", "raw-"+certID)
+
 	return certID, nil
 }
 
 func (c *CertificateManager) Delete(certID string) {
+
+	if len(certID) >= sha256.Size*2 {
```
Comment (buger) [inline certs/manager.go]: Can we remove this somehow? Maybe move this logic from CertificateManager to Certificate API handler level, where we can have access to the orgID?

File: gateway/oauth_manager.go [hunk 2]
```diff
@@ -688,10 +712,28 @@ func (r *RedisOsinStorageInterface) SetClient(id string, client osin.Client, ign
...
+		for _, v := range keys {
+			if strings.Contains(v, key) {
+				r.store.RemoveFromList(indexKey, v)
+			}
+		}
+	}
+	// append to oauth client index list
+	r.store.AppendToSet(indexKey, key)
+
+	log.Debug("Storing copy in set")
```
Comment (buger) [inline gateway/oauth_manager.go]: Remove debug logs

File: gateway/oauth_manager.go [hunk 3]
```diff
@@ -688,10 +712,28 @@ func (r *RedisOsinStorageInterface) SetClient(id string, client osin.Client, ign
...
+	log.Debug("Storing key in oauth clients index list")
+	indexKey := prefixClientIndexList + orgID
+	//check if the indexKey exists
+	exists, err := r.store.Exists(indexKey)
+	if err != nil {
+		return err
+	}
+	// if it exists, delete the key to avoid duplicity in the client index list
+	if exists {
+		keys, _ := r.store.GetListRange(indexKey, 0, -1)
```
Comment (buger) [inline gateway/oauth_manager.go]: Can we instead of retrieving the list all the time, we can just do the "lrem" of this value?

Comment (tbuchaillot) [reply gateway/oauth_manager.go]: solved

**Pull request merged**